enzyme/Enzyme/InstructionBatcher.h

```diff
@@ -0,0 +1,242 @@
+      ValueToValueMapTy vmap;
+      Instruction *new_inst = placeholder->clone();
+      vmap[placeholder] = new_inst;
+
+      for (unsigned j = 0; j < inst.getNumOperands(); ++j) {
+        Value *op = inst.getOperand(j);
+        Value *new_op;
+        if (isa<Constant>(op)) {
+          new_op = op;
+        } else if (toVectorize.count(op) != 0) {
+          new_op = vectorizedValues[op][i];
+        } else {
+          new_op = originalToNewFn[op];
+        }
+        vmap[placeholder->getOperand(j)] = new_op;
+      }
+
+      if (placeholders.size() == width) {
+        Instruction *placeholder = cast<Instruction>(placeholders[i]);
+        ReplaceInstWithInst(placeholder, new_inst);
+        vectorizedValues[&inst][i] = new_inst;
+      } else if (placeholders.size() == 1) {
+        Instruction *insertionPoint = placeholder->getNextNode()
+                                          ? placeholder->getNextNode()
+                                          : placeholder;
+        IRBuilder<> Builder2(insertionPoint);
+        Builder2.SetCurrentDebugLocation(DebugLoc());
+        Builder2.Insert(new_inst);
+        vectorizedValues[&inst].push_back(new_inst);
+      } else {
+        llvm_unreachable("Unexpected number of values in mapping");
+      }
+
+      RemapInstruction(new_inst, vmap, RF_NoModuleLevelChanges);
+
+      if (!inst.getType()->isVoidTy() && inst.hasName())
+        new_inst->setName(inst.getName() + Twine(i));
+    }
+  }
+
+  void visitPHINode(PHINode &phi) {
+    PHINode *placeholder = cast<PHINode>(vectorizedValues[&phi][0]);
+
+    for (unsigned i = 1; i < width; ++i) {
+      ValueToValueMapTy vmap;
+      Instruction *new_phi = placeholder->clone();
+      vmap[placeholder] = new_phi;
+
+      for (unsigned j = 0; j < phi.getNumIncomingValues(); ++j) {
+        Value *orig_block = phi.getIncomingBlock(j);
+        BasicBlock *new_block = cast<BasicBlock>(originalToNewFn[orig_block]);
+        Value *orig_val = phi.getIncomingValue(j);
+        Value *new_val;
+        if (isa<Constant>(orig_val)) {
+          new_val = orig_val;
+        } else {
+          new_val = vectorizedValues[orig_val][i];
+        }
+        vmap[placeholder->getIncomingValue(j)] = new_val;
+        vmap[new_block] = new_block;
+      }
+
+      RemapInstruction(new_phi, vmap, RF_NoModuleLevelChanges);
+      Instruction *placeholder = cast<Instruction>(vectorizedValues[&phi][i]);
+      ReplaceInstWithInst(placeholder, new_phi);
+      new_phi->setName(phi.getName());
+      vectorizedValues[&phi][i] = new_phi;
+    }
+  }
+
+  void visitSwitchInst(llvm::SwitchInst &inst) {
+    EmitFailure("SwitchConditionCannotBeVectorized", inst.getDebugLoc(), &inst,
+                "switch conditions have to be scalar values", inst);
+    llvm_unreachable("vectorized control flow is not allowed");
+  }
+
+  void visitBranchInst(llvm::BranchInst &branch) {
+    EmitFailure("BranchConditionCannotBeVectorized", branch.getDebugLoc(),
```

Comment:
At one point it would be cool to have as a fallback, check if these are all the same -- and if not go to an error.